src/main/java/com/atlassian/db/replica/internal/ReplicaStatement.java

```diff
@@ -513,14 +589,23 @@ void recordWriteAfterQueryExecution() throws SQLException {
         }
     }
 
-    public Statement getReadStatement(String sql) {
+    public Statement getReadStatement(String sql, QueryContext.Builder contextBuilder) {
         isWriteOperation = isFunctionCall(sql) || isUpdate(sql) || isDelete(sql);
+        if (isWriteOperation) {
+            contextBuilder.reason(WRITE_OPERATION).runOnMain(true);
+            return prepareWriteStatement();
+        }
+        if (connectionProvider.getState().equals(MAIN)) {
+            contextBuilder.reason(MAIN_CONNECTION_REUSE).runOnMain(true);
+            return prepareWriteStatement();
+        }
 
-        // if write connection is already initialized, but the current statement is null
-        // we should use a write statement, regardless of the fact that readStatement has been called.
-        if (connectionProvider.getState().equals(MAIN) || isWriteOperation || isSelectForUpdate(sql)) {
+        if (isSelectForUpdate(sql)) {
+            contextBuilder.reason(LOCK).runOnMain(true);
```

Comment:
Does it mean that `SELECT ... FOR UPDATE` locks the row/table/whatever and that's why replica is not ok? Or is the `UPDATE` itself the reason? Or both? Maybe we need `FOR_UPDATE` reason?

Reply:
`SELECT ... FOR UPDATE` is [a form of lock](https://www.postgresql.org/docs/9.4/explicit-locking.html). I don't think we need a better granularity. There's SQL in the context to determine what kind of lock causes it.